staging/kos/cmd/attachment-tput-driver/schedule.go

```diff
@@ -0,0 +1,84 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"math/rand"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// A threadOpsSchedule is the schedule of operations on attachments for a single
+// thread. It consists of a list `dt0,..,dtn` of time intervals in nanoseconds.
+// The ith operation that the thread must perform must happen at `t0 + dti`,
+// where t0 is an arbitrary start time for operations on attachments.
+// "Arbitrary start time" means that the start time is not hardcoded in the
+// schedule, and it is up to the schedule user to provide it. This is done
+// because computing a schedule might take time; if the schedule was dependent
+// on a t0, by the time it was computed t0 + dti might already have elapsed.
+// Compute the schedule first, and then pick a t0.
+type threadOpsSchedule []int64
+
+// Supported distributions of the operations on attachments.
+const (
+	steadyDistribution  = "steady"
+	poissonDistribution = "poisson"
+)
+
+// newOpsSchedule returns a schedule for the operations on attachments. The
+// returned schedule satisfies the `opsDistribution` and `opsPeriod` parameters.
+func newOpsSchedule(opsDistribution string, opsPeriod float64, totalOps, numThreads uint64) map[uint64]threadOpsSchedule {
+	threadIDToOpsSchedule := make(map[uint64]threadOpsSchedule, numThreads)
+
+	// For each operation on attachments, compute a dt such that the operation
+	// should take place at `t0 + dt` where t0 is the arbitrary start time for
+	// operations managed outside of this function.
+	var dtNanos int64
+	for i := uint64(0); i < totalOps; i++ {
+		if opsDistribution == poissonDistribution {
+			// The time between an op and the next one is given by the
+			// exponential distribution with rate `1/opsPeriod`.
+			dtNanos += int64(float64(time.Second) * (opsPeriod * rand.ExpFloat64()))
+		} else {
+			// The ops on attachments happen with a constant period.
+			dtNanos += int64(float64(time.Second) * opsPeriod)
+		}
+		if dtNanos < 0 {
+			// Overflow! This should never happen, because at every iteration
+			// dtNanos is less than or (approximately) equal to the time of the
+			// whole throughput driver run, and an int64 is able to accommodate
+			// runs that last longer than 20 years. We check anyway for safety.
+			glog.Error("Encountered an overflow while computing timing of operations on attachments; lower number of attachments to create or increase ops rate to resolve.")
+			os.Exit(100)
+		}
+		thd := i % numThreads
+		if i < numThreads {
+			threadIDToOpsSchedule[thd] = make(threadOpsSchedule, 0, 1+totalOps/numThreads)
+		}
+		threadIDToOpsSchedule[thd] = append(threadIDToOpsSchedule[thd], dtNanos)
+	}
+
+	//! The memory footprint of the schedule is theta(N) where N is the total
```

Comment:
This comment belongs somewhere more prominent than this.

Reply:
Also, it would not be bad to define a type for a whole schedule.  And that would be a good place to put this comment.

Reply:
49677ba removes the per-thread schedule and represents a global schedule as a single slice (properly named with a type alias). The comment you are talking about is moved as you suggest.